### PR TITLE
Changes to get transformers Gemma working, as well as bug fixes

### PR DESCRIPTION
**Pull request by adconner**
Some additions to get Gemma-2b working, as implemented by transformers, as well as some bug fixes in the in-place versions of sum, mean, etc.

One divergence with torch semantics with this change is I make no attempt to support the `torch.no_grad` context (arising as the implementation of `torch._C._set_grad_enabled`). In the context of Gemma it was used for pure efficiency reasons to mark tensors independent of the input; this optimization is unneeded in the jax autograd context. I am supposing the heuristic that such efficiency is the typical reason for this context, and torch tensors will usually be explicitly `detach`ed when this is required for the correctness of the model (already supported by torch2jax). If we wish to support the `no_grad` context, it can be done by tracking this state ourselves and wrapping the result of all operators in `stop_gradient` if on.

### CODE REVIEW

**Comment by samuela** [comment]
hi @adconner, thanks for submitting this! it would be super cool to get gemma working. there's some existing, in-flight PRs to be aware of for some of these issues: 
- #21 adds support for `torch.no_grad`
- #19 adds `out` kwarg support and should help reduce the number of changes needed in this PR

i propose that we get those merged first since there already in the works and then rebase from there

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -1168,12 +1196,12 @@ def custom_new(cls, *args, **kwargs):
...
 def t2j_function(f):
   def f_jax(*args, rng=None):
-    torch_args = jax.tree_util.tree_map(Torchish, args)
+    torch_args = jax.tree.map(Torchish, args)
     with override_Tensor_constructor():
       with RngPooperContext(None if rng is None else RngPooper(rng)):
         with TorchishMode():
           out = f(*torch_args)
-    return out.value
+    return jax.tree.map(lambda e: e.value, out)
```
[the remark this answers is not in the document]
Comment (adconner) [reply torch2jax/__init__.py]: You're exactly right to use torch pytree tree_map. I wasn't aware of this functionality but it handles the transformer output (where before I was wrapping Gemma to give a jax pytree output).

[the remark this answers is not in the document]
Comment (samuela) [reply torch2jax/__init__.py]: some relevant discussion in https://github.com/samuela/torch2jax/pull/19#discussion_r2276293195

i'm fine doing torch tree_map for now, with a comment explaining this situation and giving an example of why it's necessary. but i'm also concerned that it may not be a complete solution. the contract that torch2jax maintains is that `t2j(f)` accepts native jax things as input and returns native jax things as output. if we end up returning one of these pytorch dataclasses without registering it with jax we will be breaking that contract. registering the types with jax ought to solve the issue.

alternatively we could require that all functions input to `t2j` output some kind of native jax pytree type and users should wrap or rewrite code until that is true, as @adconner seems to already be doing with gemma

[the remark this answers is not in the document]
Comment (samuela) [reply torch2jax/__init__.py]: That's good to hear. Ok, let's go ahead and `jax.tree_util.register_dataclass` the few pytorch dataclasses that we care about (iiuc there are only a small handful) and then we can use `jax.tree.map` everywhere. If we run into issues, we have the pytorch tree_map as a fallback and can mark the register_dataclass as a TODO

**Comment by adconner** [comment]
I've rebased @samuela

**Comment by samuela** [comment]
just ooc gemma is already implemented in jax, no? https://github.com/google-deepmind/gemma just out of curiosity what's your use case look like?

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -1034,7 +1047,9 @@ def prelu(input: Torchish, weight: Torchish):
 
 
 @implements(torch.nn.functional.scaled_dot_product_attention)
-def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None):
+def scaled_dot_product_attention(
+  query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None, enable_gqa=False
```
[the remark this answers is not in the document]
Comment (adconner) [reply torch2jax/__init__.py:1051]: This supports `enable_gqa=True` also. In jax its just implicit in the shapes of the input. If you like I could make the jax code throw if the shape assumptions are violated when `False`, but this would mean the original torch model didn't work under torch.

**Comment by adconner** [comment]
My use case involves possibly collaborating with torch users in the future, and potentially wanting to run experiments with other models aside from gemma.

File: torch2jax/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adconner) [reply torch2jax/__init__.py:1051]: If you like you could add the following assertion, but it would only fire if the model already didnt work under torch. Do you want such assertions in torch2jax?
```
  assert enable_gqa or Q.shape[-3] == V.shape[-3], "enable_gqa=False and number of Q and V heads disagree"
```

[the remark this answers is not in the document]
Comment (adconner) [reply torch2jax/__init__.py:1051]: What I am saying is that gqa is implemented by jax and by us. It is used in Gemma, which I am using

[the remark this answers is not in the document]
Comment (samuela) [reply torch2jax/__init__.py:1051]: that may be true but according to your current diff the `enable_gqa` argument is unused. if you want to add support for gqa we'll need to add that to the test suite.

[the remark this answers is not in the document]
Comment (samuela) [reply torch2jax/__init__.py:1051]: i see now that https://docs.jax.dev/en/latest/_autosummary/jax.nn.dot_product_attention.html automatically selects gqa based on the sizes of the inputs. in order to enable this let's add a comment explaining why this argument is unused and add tests for enable_gqa

[the remark this answers is not in the document]
Comment (adconner) [reply torch2jax/__init__.py:1051]: done

**Review by samuela — APPROVED** [review]
lgtm, thanks @adconner !
